### PR TITLE
Referenced entities

### LINKED ISSUE #239: confusing error message on trying to create rule with shared action
Trying this in the CLI

```
$ wsk rule create ifGitThenEcho myGitTrigger /whisk.system/samples/echo
error: The request content was malformed:
entity name malformed (code 1385580)
```

In `wsk rule create name trigger action` trigger or action can't be private
User needs to copy the action to his space like

```
$ wsk action create echo --copy /whisk.system/samples/echo
ok: created action echo
```

CLI error should be more clear about this and the [docs](https://github.com/openwhisk/openwhisk/blob/master/docs/triggers_rules.md#using-rules-to-associate-triggers-and-actions) should be updated to reflect this

### PR DESCRIPTION
**Pull request by rabbah**
Main features of this pr:
1. entitlement for sequences (will need more testing)
2. entitlement for rules with actions in packages (will need more tests, fixes #239)
3. fixes some gaps in the package entitlement (added several new tests)

The first 5 commits are rebased from other prs for completeness.

### CODE REVIEW

**Comment by rabbah** [comment]
pg2/5​61  approved.

File: core/controller/src/main/scala/whisk/core/controller/AuthorizedRouteDispatcher.scala [hunk 1]
```diff
@@ -66,44 +66,20 @@ trait BasicAuthorizedRouteProvider extends Directives with Logging {
...
-            case Failure(t) if recover.isDefined => recover.get(t)
-            case Failure(r: RejectRequest)       => terminate(r.code, r.message)
-            case Failure(t)                      => terminate(InternalServerError, t.getMessage)
+    protected def handleEntitlementFailure(failure: Try[Boolean])(
+        implicit transid: TransactionId): RequestContext => Unit = {
+        failure match {
+            case Success(false)            => terminate(Forbidden)
+            case Failure(r: RejectRequest) => terminate(r.code, r.message)
+            case Failure(t)                => terminate(InternalServerError, t.getMessage)
+            case _ /*Success(true)*/       => terminate(InternalServerError, "Should not reach here.")
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/AuthorizedRouteDispatcher.scala]: Sanity question: Do we surface the message of "Internal Server Error"s?

File: core/controller/src/main/scala/whisk/core/entitlement/RateThrottler.scala [hunk 2]
```diff
@@ -43,31 +43,33 @@ class RateThrottler(maxPerMinute: Int) extends Logging {
...
-            }
+            //if (curHour != lastHour) {
+            //    lastHour = curHour
+            //    lastHourCount = 0
+            //}
         }
 
         private def getCurrentMinute = System.currentTimeMillis / (60 * 1000)
-        private def getCurrentHour = System.currentTimeMillis / (3600 * 1000)
+        //private def getCurrentHour = System.currentTimeMillis / (3600 * 1000)
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/entitlement/RateThrottler.scala]: Remove all the commented lines? Why keep them in?

File: core/controller/src/main/scala/whisk/core/entitlement/RateThrottler.scala [hunk 3]
```diff
@@ -28,9 +28,9 @@ import akka.event.Logging.InfoLevel
...
  * For now, we throttle only at a 1-minute granularity.
  */
-class RateThrottler(maxPerMinute: Int) extends Logging {
+class RateThrottler(description: String, maxPerMinute: Int) extends Logging {
 
-    info(this, s"maxPerMinute = $maxPerMinute")
+    info(this, s"$description: maxPerMinute = $maxPerMinute")
 
     // Parameters
     private val exemptSubject = Set[Subject]() // We exempt no one.
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/entitlement/RateThrottler.scala]: While we're on it, remove this line?

**Comment by rabbah** [comment]
I'll rebase and restage this pull request into smaller commits.

**Pull request closed without merging**